src/tool_homedir.c

```diff
@@ -45,14 +52,50 @@ static char *GetEnv(const char *variable)
 }
 
 /* return the home directory of the current user as an allocated string */
-char *homedir(void)
+
+/*
+ * The original logic found a home dir to use (by checking a range of
+ * environment variables and last using getpwuid) and returned that for the
+ * parent to use.
+ *
+ * With the XDG_CONFIG_HOME support (added much later than the other), this
+ * variable is treated differently in order to not ruin existing installations
+ * even if this environment variable is set. If this variable is set, and a
+ * file name is set to check, then only if that file name exists in that
+ * directory will it be returned as a "home directory".
+ *
+ * 1. use CURL_HOME if set
+ * 2. use XDG_CONFIG_HOME if set and fname is present
+ * 3. use HOME if set
+ * 4. Non-windows: use getpwuid
+ * 5. Windows: use APPDATA if set
+ * 6. Windows: use "USERPROFILE\Application Data" is set
+ */
+
+char *homedir(const char *fname)
 {
   char *home;
 
   home = GetEnv("CURL_HOME");
   if(home)
     return home;
 
+  if(fname) {
+    home = GetEnv("XDG_CONFIG_HOME");
+    if(home) {
+      char *c = curl_maprintf("%s" DIR_CHAR "%s", home, fname);
+      if(c) {
+        int fd = open(c, O_RDONLY);
+        bool works = (fd >= 0);
+        close(fd);
+        curl_free(c);
+        if(works)
+          return home;
```

Comment:
```suggestion
        if(fd >= 0) {
          close(fd);
          return home;
        }
```
It is clear what fd >=0 means I don't think you need `works` var and also close(-1) could be a problem on some platforms.

Reply:
agreed!